rocAL/rocAL/source/video_read_and_decode.cpp

```diff
@@ -0,0 +1,293 @@
+    _video_path.resize(_sequence_count);
+    for (size_t i = 0; i < _sequence_count; i++)
+    {
+        _compressed_buff[i].resize(MAX_COMPRESSED_SIZE);
+        _start_frame[i] = _reader->read(_compressed_buff[i].data(), fsize);
+        _video_path[i] = _reader->id();
+        _reader->close();
+        _decompressed_buff_ptrs[i] = buff + (i * image_size * _sequence_length);
+
+        // Check if the video file is already initialized otherwise use an existing decoder instance to initialize the video 
+        // std::cerr << "\nThe source video is " << _video_path[i] << " MAP : "<<_video_file_name_map.find(_video_path[i])->second._video_map_idx << "\tThe start index is : " << _start_frame[i] << "\n";
+        std::map<std::string, video_map>::iterator itr = _video_file_name_map.find(_video_path[i]);
+        if (itr->second._is_decoder_instance == false)
+        {
+            std::map<std::string, video_map>::iterator temp_itr;
+            for (temp_itr = _video_file_name_map.begin(); temp_itr != _video_file_name_map.end(); ++temp_itr)
+            {
+                if (temp_itr->second._is_decoder_instance == true)
+                {
+                    int video_idx = temp_itr->second._video_map_idx;
+                    if ((std::count(_sequence_video_idx.begin(), _sequence_video_idx.end(), video_idx) >= 1) && (_sequence_video_idx.size() != 0))
+                        continue;
+                    std::vector<std::string> substrings;
+                    char delim = '#';
+                    substring_extraction(itr->first, delim, substrings);
+                    if (_video_decoder[video_idx]->Initialize(substrings[1].c_str()) == VideoDecoder::Status::OK)
+                    {
+                        itr->second._video_map_idx = video_idx;
+                        itr->second._is_decoder_instance = true;
+                    }
+                    temp_itr->second._is_decoder_instance = false;
+                    break;
+                }
+            }
+        }
+        if (itr->second._is_decoder_instance == false)
+            continue;
+        _sequence_video_idx.push_back(itr->second._video_map_idx);
+
+        // Check if the sequences are from same or different video file 
+        video_index.push_back(_video_file_name_map[_video_path[i]]._video_map_idx);
+        if (std::count(video_index.begin(), video_index.end(), video_index[i]) > 1)
+        {
+            sequential_decode_sequences.push_back(i);
+        }
+        else
+        {
+            parallel_decode_sequences.push_back(i);
+            parallel_sequence_count++;
+        }
+    }
+    if (parallel_sequence_count == 1)
+    {
+        sequential_decode_sequences.push_back(parallel_decode_sequences.back());
+        parallel_decode_sequences.clear();
+    }
+
+    _file_load_time.end(); // Debug timing
+
+    _decode_time.start(); // Debug timing
+
+    for (size_t i = 0; i < sequential_decode_sequences.size(); i++)
```

Comment:
better to start the parallel decoder threads before sequential decoding, so they can both work in parallel.
In other words, move the sequential decode loop after line #262 and before join

Reply:
We have already tried this method, since there is a possibility that the when the sequences are being decoded in parallel, the sequential decoding may try to access a sequence of the same video file as the one being decoded in parallel, hence it throws Segmentation fault. So we will not be able to do this way.